libraries/AP_Compass/AP_Compass.cpp

```diff
@@ -472,12 +456,29 @@ Compass::init()
         // detect available backends. Only called once
         _detect_backends();
     }
-    if (_compass_count != 0) {
-        // get initial health status
-        hal.scheduler->delay(100);
+    initialise_health_status();
+    return true;
+}
+
+void Compass::initialise_health_status(void)
+{
+    if (_compass_count == 0) {
+        return;
+    }
+
+    // get initial health status
+    for (uint8_t attempts=0; attempts<10; attempts++) {
         read();
+        for (uint8_t i=0; i < COMPASS_MAX_INSTANCES; i++) {
+            if (!_state[i].healthy) {
+                hal.scheduler->delay(10);
+                continue;
```

Comment:
it doesn't seems right ... the continue will only affect the second for loop and thus do nothing, right ?

Reply:
I had to look it up but, I think @khancyr is right.